### DISCUSSION TITLE
How to get the values from a Sales Invoice products table?

### DISCUSSION THREAD

**Discussion by nihilne** [YetiForceCompany/YetiForce#41] [How to] [answered]
Hello,

I've created a custom workflow that reads the values of fields in a Sales Invoice and sends them to a third-party API. So far, I've managed to read values of normal fields using `$recordModel->get('subject')`.

How do I go about reading a products table/advanced block, and adding the products to an array using a loop? Is there maybe an existing function that does this or a variable that contains all of the products?

I've tried reading the Documentation but I couldn't find anything that could solve my problem.

**Comment by nihilne** [comment]
I might have figured something out, although any improvements or recommendations are highly appreciated

```php
<?php
class FInvoiceWorkflow {
    public function getInvoiceAndPost(Vtiger_Record_Model $recordModel) {
        \App\Log::trace('Entering getInvoiceAndPost');

        $productsArray = [];
        $inventoryData = $recordModel->getInventoryData();
        foreach ($inventoryData as $product) {
            $productRecordModel = Vtiger_Record_Model::getInstanceById($product['name'], 'Products');
            $productName = $productRecordModel->get('name');
            /**Other fields here */
            $singleProduct = [
                'lineCode' => $productName
            ];
            array_push($productsArray, $singleProduct); /**Push a single product into the products array */
        }
        \App\Log::trace('getInvoiceAndPost Done'); 
    }
}
?>
```

**Comment by nihilne** [comment]
After some more experimenting, I removed the `$productsRecordModel` variable, and I instead did `$productName = $product['name']`

**Comment by rskrzypczak** [comment, marked as answer]
You have two errors here:
- in the advanced block there doesn't have to be only products, so it's best to remove the second argument when initializing the record model 
- if you already have a product object, you refer to fields from its module, so you should give `productname` instead of `name`
![image](https://github.com/user-attachments/assets/7062df8c-ccc9-45cc-9c47-3d9043f18621)

**Reply by nihilne** [reply]
Aaah, yeah I see how that works now. Thank you very much!
